geoana/em/static/halfspace.py

```diff
@@ -0,0 +1,628 @@
+
+        The ``DipoleHalfSpace`` class is used to analytically compute the
+        potentials, current densities and electric fields within a halfspace due to a dipole source.
+        Surface is assumed to be at z=0.
+
+        Parameters
+        ----------
+        current : float
+            Electrical current in the point current (A). Default is 1A.
+        rho : float
+            Resistivity in the point current (:math:`\\Omega \\cdot m`).
+        locations : array_like
+            Locations of the M and N electrodes (m).
+        """
+
+    def __init__(self, rho, locations, current=1.0):
+
+        _primary = PointCurrentHalfSpace(rho, current=1.0, location=locations[0])
+        _secondary = PointCurrentHalfSpace(rho, current=-1.0, location=locations[1])
+        self._primary = _primary
+        self._secondary = _secondary
+
+        self.current = current
+        self.rho = rho
+        self.locations = locations
+
+    @property
+    def current(self):
+        """Current in the point current in Amps.
+
+        Returns
+        -------
+        float
+            Current in the point current in Amps.
+        """
+        return self._current
+
+    @current.setter
+    def current(self, value):
+
+        try:
+            value = float(value)
+        except:
+            raise TypeError(f"current must be a number, got {type(value)}")
+
+        self._current = value
+        self._primary.current = value
+        self._secondary.current = -value
+
+    @property
+    def rho(self):
+        """Resistivity in the point current in :math:`\\Omega \\cdot m`.
+
+        Returns
+        -------
+        float
+            Resistivity in the point current in :math:`\\Omega \\cdot m`.
+        """
+        return self._rho
+
+    @rho.setter
+    def rho(self, value):
+
+        try:
+            value = float(value)
+        except:
+            raise TypeError(f"current must be a number, got {type(value)}")
+
+        if value <= 0.0:
+            raise ValueError("current must be greater than 0")
+
+        self._rho = value
+        self._primary.rho = value
+        self._secondary.rho = value
+
+    @property
+    def locations(self):
+        """Locations of the two voltage electrodes.
+
+        Returns
+        -------
+        (2, 3) numpy.ndarray of float
+            Locations of the two voltage electrodes.
+        """
+        return self._locations
+
+    @locations.setter
+    def locations(self, vec):
+
+        try:
+            vec = np.atleast_1d(vec).astype(float)
+        except:
+            raise TypeError(f"location must be array_like, got {type(vec)}")
+
+        if np.any(vec[..., -1] > 0):
+            raise ValueError(
+                f"z value must be less than or equal to 0 in a halfspace, got {(vec[..., -1])}"
+            )
+
+        self._locations = vec
+        self._primary.location = vec[0]
+        self._secondary.location = vec[1]
+
+    def potential(self, xyz):
```

Comment:
for these functions here, I was imagining there being an optional second list of xyz locations, so the function signature would be:
`def potential(self, xyz_m, xyz_n=None)`

Then to evaluate it just do a conditional check and only evaluate and subtract it at the second set of locations `if xyz_n is not None`

and so on for the other evaluations